### PR TITLE
Export setting Save logic

### PR DESCRIPTION
**Pull request by DhaaraniCIT**

### CODE REVIEW

File: src/app/core/models/misc/tracking.model.ts [hunk 1]
```diff
@@ -2,14 +2,15 @@ import { ProgressPhase } from "../enum/enum.model";
...
 
 export type UpdateEventAdditionalProperty = {
     phase: ProgressPhase,
-    oldState: QBDAdvancedSettingsGet | QBDExportSettingGet | QBDFieldMappingGet,
-    newState: QBDAdvancedSettingsGet | QBDExportSettingGet | QBDFieldMappingGet
+    oldState: QBDAdvancedSettingsGet | QBDExportSettingGet | QBDFieldMappingGet | Sage300ExportSettingGet,
+    newState: QBDAdvancedSettingsGet | QBDExportSettingGet | QBDFieldMappingGet | Sage300ExportSettingGet
   };
 
   export type UpdateIntacctEventAdditionalProperty = {
```
Comment (DhaaraniCIT) [inline src/app/core/models/misc/tracking.model.ts:16]: @ashwin1111 can we reuse `UpdateEventAdditionalProperty` as common one and delete `UpdateIntacctEventAdditionalProperty`

File: src/app/core/services/integration/tracking.service.ts [hunk 2]
```diff
@@ -82,15 +84,15 @@ export class TrackingService {
     this.eventTrack(`Time Spent on ${page} page`, {durationInSeconds: differenceInMs / 1000});
   }
 
-  onOnboardingStepCompletion(eventName: QBDOnboardingState, stepNumber: number, additionalProperties: QBDExportSettingPost | QBDFieldMappingPost | QBDAdvancedSettingsPost | void): void {
+  onOnboardingStepCompletion(eventName: QBDOnboardingState | Sage300OnboardingModule, stepNumber: number, additionalProperties: QBDExportSettingPost | QBDFieldMappingPost | QBDAdvancedSettingsPost | void | Sage300ExportSettingPost): void {
     this.eventTrack(`Step ${stepNumber} completed: ${eventName}`, additionalProperties);
   }
 
   integrationsOnboardingCompletion(eventName: IntacctOnboardingState, stepNumber: number, additionalProperties: LocationEntityPost | ExportSettingPost | ImportSettingPost | AdvancedSettingsPost | void): void {
```
Comment (DhaaraniCIT) [inline src/app/core/services/integration/tracking.service.ts:91]: can we make it common?

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 3]
```diff
@@ -69,8 +78,66 @@ export class Sage300ExportSettingsComponent implements OnInit {
     this.helperService.importAttributes(isRefresh);
   }
 
-  save() {
-    // Will be added here soon
+  exportTypeWatcher() {
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:81]: rule = [
  {
    'formController': 'reimbursableExportType',
    'requiredValues': {
      'DIRECT_COST': ['defaultReimbursableCCCAccountName']
    }
  },
  {
    'formController': 'cccExportType',
    'requiredValues': {
      'DIRECT_COST': ['defaultCreditCardCCCAccountName'],
      'PURCHASE_INVOICE': ['defaultCreditCardCCCAccountName']
    }
  }
]

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 4]
```diff
@@ -69,8 +78,66 @@ export class Sage300ExportSettingsComponent implements OnInit {
...
+  }
+
+  private constructPayloadAndSave(): void {
+    this.isSaveInProgress = true;
+    const exportSettingPayload = ExportSettingModel.createExportSettingPayload(this.exportSettingForm);
+    this.exportSettingService.postExportSettings(exportSettingPayload).subscribe((exportSettingResponse: Sage300ExportSettingGet) => {
+      this.isSaveInProgress = false;
+      this.toastService.displayToastMessage(ToastSeverity.SUCCESS, 'Export settings saved successfully');
+      this.trackingService.trackTimeSpent(Page.EXPORT_SETTING_SAGE300, this.sessionStartTime);
+      this.isSaveInProgress = false;
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:111]: repeated twice, remove

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 5]
```diff
@@ -176,8 +171,8 @@ export class Sage300ExportSettingsComponent implements OnInit {
         this.exportSettingForm = ExportSettingModel.mapAPIResponseToFormGroup();
       }
       this.addFormValidator();
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:173]: you can assign response[0] to export settings directly and call mapAPI()
```
this.exportSettings = response[0];
this.exportSettingForm = ExportSettingModel.mapAPIResponseToFormGroup(this.exportSettings);

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 6]
```diff
@@ -165,6 +140,26 @@ export class Sage300ExportSettingsComponent implements OnInit {
 
   private setupPage(): void {
     this.isOnboarding = this.router.url.includes('onboarding');
+    const validatorRule: ExportSettingValidatorRule = {
+      'reimbursableExpense': ['reimbursableExportType', 'reimbursableExportGroup', 'reimbursableExportDate', 'reimbursableExpenseState'],
+      'creditCardExpense': ['cccExportType', 'cccExportGroup', 'cccExportDate', 'cccExpenseState']
+    };
+
+    const exportSettingValidatorRule: ExportSettingExportTyleValidatorRule[] = [
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:148]: exportModuleRule

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 7]
```diff
@@ -165,6 +140,26 @@ export class Sage300ExportSettingsComponent implements OnInit {
 
   private setupPage(): void {
     this.isOnboarding = this.router.url.includes('onboarding');
+    const validatorRule: ExportSettingValidatorRule = {
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:143]: exportSettingValidatorRule

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts (hunk 6 above)
Comment (ashwin1111) [reply src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:148]: ExportModuleRule[] for type

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 8]
```diff
@@ -165,6 +140,26 @@ export class Sage300ExportSettingsComponent implements OnInit {
...
+      'reimbursableExpense': ['reimbursableExportType', 'reimbursableExportGroup', 'reimbursableExportDate', 'reimbursableExpenseState'],
+      'creditCardExpense': ['cccExportType', 'cccExportGroup', 'cccExportDate', 'cccExpenseState']
+    };
+
+    const exportSettingValidatorRule: ExportSettingExportTyleValidatorRule[] = [
+      {
+        'formController': 'reimbursableExportType',
+        'expectedValue': {
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:151]: let's call it `requiredValue`

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 9]
```diff
@@ -176,8 +171,8 @@ export class Sage300ExportSettingsComponent implements OnInit {
         this.exportSettingForm = ExportSettingModel.mapAPIResponseToFormGroup();
       }
       this.addFormValidator();
-      this.helper.setCustomValidatorsAndWatchers(this.validatorRule, this.exportSettingForm);
-      this.helper.setCustomExportTypeValidatoresAndWatchers(this.exportSettingValidatorRule, this.exportSettingForm);
+      this.helper.setCustomValidatorsAndWatchers(validatorRule, this.exportSettingForm);
+      this.helper.setCustomExportTypeValidatoresAndWatchers(exportSettingValidatorRule, this.exportSettingForm);
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:175]: name the functions similar to variable suggestions added above

**Review by ashwin1111 — APPROVED** [review]

**Pull request merged**